Reduce ODR violations when using Objective-C ARC

geoffreygaren opened the pull request:
#### 0164d04ec7e4eadc7c63b2ff2689a42e660ef471
<pre>
Reduce ODR violations when using Objective-C ARC
<a href="https://bugs.webkit.org/show_bug.cgi?id=288032">https://bugs.webkit.org/show_bug.cgi?id=288032</a>
<a href="https://rdar.apple.com/145180401">rdar://145180401</a>

Reviewed by Ryosuke Niwa and David Kilzer.

When an ARC file and a non-ARC file both #include the same header, for each
inline function in the header, the compiler emits two versions — the ARC
version and the non-ARC version.

This includes implicitly generated inline functions like constructors,
assignment operators, and destructors.

Technically, having two versions of an inline function is always a violation of
C++&apos;s &quot;One Definition Rule&quot;, and is always undef, even if both functions have
similar behavior (&quot;...each definition of D shall consist of the same sequence
of tokens... implicit calls... shall refer to the same function...&quot;)

In practice, you only get a leak or a crash when the ARC version and the non-ARC
version have different behavior, for example:
    * when the function initializes, modifies, or destroys a stored property
    * when the function returns a pointer to an ARC type
    * when the function declares NS_RETURNS_RETAINED or NS_RELEASES_ARGUMENT

Example: <a href="https://rdar.apple.com/73620184">rdar://73620184</a>.

I haven’t figured out how to solve this yet.

For now, this patch extends our existing workaround to make mistakes less
likely as we roll out smart pointers across WebKit. Specifically, it gives
makeBlockPtr, RetainPtr, and OSObjectPtr different names in ARC and non-ARC
files. This ensures that their inline functions can’t conflict with each other
across files.

As a consequence, it is no longer possible to pass a RetainPtr argument
from an ARC file to a non-ARC file, or vice-versa.

* Source/WTF/wtf/BlockPtr.h:
* Source/WTF/wtf/Forward.h:
* Source/WTF/wtf/OSObjectPtr.h:
* Source/WTF/wtf/RetainPtr.h:
* Source/WTF/wtf/cocoa/WorkQueueCocoa.cpp:
(WTF::WorkQueueBase::dispatchWithQOS):
* Source/WebKit/Shared/API/APIData.h:
* Source/WebKit/Shared/Cocoa/APIDataCocoa.mm:
(API::Data::createWithoutCopying):

Canonical link: <a href="https://commits.webkit.org/290734@main">https://commits.webkit.org/290734@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fe7582e599a01ec46a319c48d4de195ff98cf2f8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90914 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10460 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45880 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95944 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41716 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92967 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10855 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18776 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/69916 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27439 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93915 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8269 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82423 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50246 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8042 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36816 "Found 60 new test failures: fast/workers/worker-page-cache.html http/tests/security/clipboard/copy-paste-html-cross-origin-iframe-across-origin.html http/tests/security/contentSecurityPolicy/block-all-mixed-content/insecure-script-in-iframe.html imported/w3c/web-platform-tests/css/css-conditional/css-supports-033.xht imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-004.html imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-005.html imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-grid-001.html imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-grid-002.html imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline-grid-003.html imported/w3c/web-platform-tests/css/css-grid/alignment/grid-align-baseline.html ... (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40837 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/83742 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78348 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37885 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97917 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/89696 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18120 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13302 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/78933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18379 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78253 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78117 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22600 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/21215 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14343 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18125 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/23474 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/112262 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17862 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/32623 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21320 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19647 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WTF/wtf/cocoa/WorkQueueCocoa.cpp:
```diff
@@ -56,14 +56,11 @@ void WorkQueueBase::dispatch(Function<void()>&& function)
 void WorkQueueBase::dispatchWithQOS(Function<void()>&& function, QOS qos)
 {
-    dispatch_block_t blockWithQOS = dispatch_block_create_with_qos_class(DISPATCH_BLOCK_ENFORCE_QOS_CLASS, Thread::dispatchQOSClass(qos), 0, makeBlockPtr([function = WTFMove(function)] () mutable {
+    OSObjectPtr blockWithQOS = dispatch_block_create_with_qos_class(DISPATCH_BLOCK_ENFORCE_QOS_CLASS, Thread::dispatchQOSClass(qos), 0, makeBlockPtr([function = WTFMove(function)] () mutable {
```
This leaks because `dispatch_block_create_with_qos_class()` has “create” in its name which means it returns a +1 retained object.

You need to use `adoptOSObject()` here, maybe with `auto`.

geoffreygaren replied:
Perhaps emphasizing how fraught it is to reason about ARC in a mixed-language / mixed-mode project, I don't think this is a leak. ARC is enabled in this file, and therefore ARC "owns" the local +1 created by "create", and will release it at end of scope. Relatedly, `retainOSObject()`, called by the `OSObjectPtr` constructor, is a no-op.

Still, good to mirror our current coding idioms. I'll fix.

ddkilzer replied:
LOL!  Yeah, it's going to be nearly impossible to review code without knowing whether each file is ARC or MRR.

(Even if you put the compiler check at the top of the file, it still means you need to click your way to the top of the file to check during reviews.)

ddkilzer replied:
I suppose this is another argument for not using `RetainPtr<>` in ARC sources, though it still has the same issue of not knowing which files have ARC enabled when reviewing.

ddkilzer reviewed (approved):
r=me if you fix the leak

geoffreygaren committed:
Reduce ODR violations when using Objective-C ARC
https://bugs.webkit.org/show_bug.cgi?id=288032
rdar://145180401

Reviewed by Ryosuke Niwa and David Kilzer.

When an ARC file and a non-ARC file both #include the same header, for each
inline function in the header, the compiler emits two versions — the ARC
version and the non-ARC version.

This includes implicitly generated inline functions like constructors,
assignment operators, and destructors.

Technically, having two versions of an inline function is always a violation of
C++'s "One Definition Rule", and is always undef, even if both functions have
similar behavior ("...each definition of D shall consist of the same sequence
of tokens... implicit calls... shall refer to the same function...")

In practice, you only get a leak or a crash when the ARC version and the non-ARC
version have different behavior, for example:
    * when the function initializes, modifies, or destroys a stored property
    * when the function returns a pointer to an ARC type
    * when the function declares NS_RETURNS_RETAINED or NS_RELEASES_ARGUMENT

Example: rdar://73620184.

I haven’t figured out how to solve this yet.

For now, this patch extends our existing workaround to make mistakes less
likely as we roll out smart pointers across WebKit. Specifically, it gives
makeBlockPtr, RetainPtr, and OSObjectPtr different names in ARC and non-ARC
files. This ensures that their inline functions can’t conflict with each other
across files.

As a consequence, it is no longer possible to pass a RetainPtr argument
from an ARC file to a non-ARC file, or vice-versa.

* Source/WTF/wtf/BlockPtr.h:
* Source/WTF/wtf/Forward.h:
* Source/WTF/wtf/OSObjectPtr.h:
* Source/WTF/wtf/RetainPtr.h:
* Source/WTF/wtf/cocoa/WorkQueueCocoa.cpp:
(WTF::WorkQueueBase::dispatchWithQOS):
* Source/WebKit/Shared/API/APIData.h:
* Source/WebKit/Shared/Cocoa/APIDataCocoa.mm:
(API::Data::createWithoutCopying):

Canonical link: https://commits.webkit.org/290734@main
